render/render.go

```diff
@@ -1,98 +1,42 @@
 package render
 
 import (
-	"bytes"
-	"errors"
-	"fmt"
-	"github.com/sasswart/gin-in-a-can/config"
-	"github.com/sasswart/gin-in-a-can/openapi"
-	"golang.org/x/text/cases"
-	"golang.org/x/text/language"
-	"os"
-	"path/filepath"
-	"strings"
+	"github.com/sasswart/gin-in-a-can/tree"
 	"text/template"
 )
 
-// Render is the main parsing and rendering steps within the render library
-func Render(config config.Config, data openapi.Traversable, templateFile string) ([]byte, error) {
-	var absoluteTemplateDirectory string
-	switch true {
-	case filepath.IsAbs(config.Generator.TemplateDirectory):
-		absoluteTemplateDirectory = filepath.Join(
-			config.Generator.TemplateDirectory,
-			config.Generator.TemplateName,
-		)
-	case filepath.IsAbs(config.ConfigFilePath):
-		absoluteTemplateDirectory = filepath.Join(
-			filepath.Dir(config.ConfigFilePath),
-			config.Generator.TemplateDirectory,
-			config.Generator.TemplateName,
-		)
-	default:
-		absoluteTemplateDirectory = filepath.Join(
-			config.WorkingDirectory,
-			filepath.Dir(config.ConfigFilePath),
-			config.Generator.TemplateDirectory,
-			config.Generator.TemplateName,
-		)
-	}
+var _ Renderer = &Base{}
 
-	var absoluteOutputFile string
-	switch true {
-	case filepath.IsAbs(config.OutputPath):
-		absoluteOutputFile = config.OutputPath
-	case filepath.IsAbs(config.ConfigFilePath):
-		absoluteOutputFile = filepath.Join(
-			filepath.Dir(config.ConfigFilePath),
-			config.OutputPath,
-		)
-	default:
-		absoluteOutputFile = filepath.Join(
-			config.WorkingDirectory,
-			filepath.Dir(config.ConfigFilePath),
-			config.OutputPath,
-		)
-	}
-	absoluteOutputFile = filepath.Join(absoluteOutputFile, data.GetOutputFile())
+type Renderer interface {
+	SanitiseName([]string) string
+	SanitiseType(n tree.NodeTraverser) string
 
-	buff := bytes.NewBuffer([]byte{})
-
-	templater := template.New(templateFile)
-
-	templater.Funcs(templateFuncMap)
-
-	parsedTemplate, err := templater.ParseGlob(fmt.Sprintf("%s/*.tmpl", absoluteTemplateDirectory))
-	if err != nil {
-		return nil, err
-	}
-
-	err = parsedTemplate.Execute(buff, data)
-	if err != nil {
-		return nil, err
-	}
-
-	fmt.Printf("Rendering %s using %s\n", data.GetOutputFile(), templateFile)
+	GetOutputFilename(n tree.NodeTraverser) string
+	SetTemplateFuncMap(*template.FuncMap)
+	GetTemplateFuncMap() template.FuncMap
+}
 
-	outputDirectory := filepath.Dir(absoluteOutputFile)
-	if _, err := os.Stat(outputDirectory); errors.Is(err, os.ErrNotExist) {
-		_ = os.MkdirAll(outputDirectory, 0755)
-	}
+// Base defines the base render object. This should be used as a compositional base for specialising it's interface
+// towards a specific use case.
+type Base struct {
+	TemplateFuncMapping *template.FuncMap
+}
 
-	err = os.WriteFile(absoluteOutputFile, buff.Bytes(), 0644)
-	if err != nil {
-		return nil, err
-	}
+func (b *Base) SanitiseName(_ []string) string {
```

Comment:
This turns a compile time interface error into a runtime panic. We do this in our generated code in order to provide an example, but I don't think that's a good idea here.

Perhaps it would be best to just let consumers satisfy the interface above themselves?

Reply:
SetChild, but yes.

Keen on your thoughts on this.

Reply:
I've removed them all ;)
As long as the overrides work, I can gain the same value I get from a complete crash by stepping through with a debugger when behaviour becomes weird